商品削除機能の実装

keeee-a opened the pull request:
# What
商品削除機能の実装
# Why
ユーザーが商品の出品の取り消しができるようにするため
# 補足
実装条件の「 出品者だけが商品情報を削除できること」は詳細ページで、出品者でなければ削除ボタンが表示されないようになっているので、出品者だけが商品情報を削除できると考えた。
# Gyazo
[商品情報の削除](https://gyazo.com/74d09f5ddb99ed9985bd4571eb5b8641)

keeee-a committed:
出品した商品を削除できるようにする。

keeee-a committed:
商品情報を削除できたかどうかの条件分岐を削除

削除できたかどうかの条件分岐をする指示がなかったため

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -36,6 +36,12 @@ def update
   end
 
+  def destroy
+    @item = Item.find(params[:id])
```
```ruby
    @item = Item.find(params[:id])
```
こちらの記述は削除し、`set_item`メソッドを利用しましょう！

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -36,6 +36,12 @@ def update
 
+  def destroy
+    @item = Item.find(params[:id])
+    @item.destroy
```
```ruby
    @item.destroy
```
destroyアクションに、削除に失敗した場合の条件分岐を記述しましょう。削除ボタンが表示されるビューをrenderさせるのが良いでしょう。
理由：予期せぬバグを防ぐため

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -36,6 +36,12 @@ def update
+  def destroy
+    @item = Item.find(params[:id])
+    @item.destroy
+    redirect_to action: :index
```
destroyアクションに、
viewの条件と同様に「出品したユーザーでないと削除できない」という条件分岐を記述しておきましょう。

理由：検証ツールを用いると削除ボタンを作成することはできてしまうので、サーバー側でも対応する必要があるため。

chisesakai reviewed (commented):
お待たせ致しました、レビュー依頼ありがとうございます。
レビュー依頼ありがとうございます。
挙動、コードともに確認致しました。
#### 3件コメントいたしましたので、ご確認をお願いいたします。

keeee-a committed:
修正

削除に失敗した場合の条件分岐
出品したユーザでないと削除できないようにする
set_itemを用いて可読性を上げる

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -1,5 +1,5 @@
 class ItemsController < ApplicationController
-  before_action :set_item, only: [:show, :edit, :update]
+  before_action :set_item, only: [:show, :edit, :update, :destroy]
   before_action :authenticate_user!, only: [:new, :edit]
```
```ruby
  before_action :authenticate_user!, only: [:new, :edit]
```
「ログアウト中のユーザー」を弾くため、こちらにも`destroy`アクションを追加しましょう！

chisesakai commented on app/controllers/items_controller.rb:
```diff
@@ -36,6 +36,16 @@ def update
   end
 
+  def destroy
+    if @item.user.id == current_user.id
```
```ruby
    if @item.user.id == current_user.id
```
条件分岐の追加、ありがとうございます！
こちら、`edit`アクションと同様、「出品者ではなければどうするか」の記述も追加しましょう。

その際、`edit`アクションと記述が重複するかと思いますので、private以下にメソッド化し,before_actionで呼び出す形式で記述してみましょう！

chisesakai commented on config/routes.rb:
```diff
@@ -2,5 +2,5 @@
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
 root to: "items#index"
-resources :items, only:[:index, :new, :create, :show, :edit, :update]
+resources :items, only:[:index, :new, :create, :show, :edit, :update, :destroy]
```
```ruby
resources :items, only:[:index, :new, :create, :show, :edit, :update, :destroy]
```
こちら、7つのアクションすべてが揃っていますので、`, only:[:index, :new, :create, :show, :edit, :update, :destroy]`の部分は削除しましょう。

chisesakai reviewed (commented):
レビュー依頼ありがとうございます。
早速の修正、ありがとうございます！
修正点確認させていただきました。どんどんより良いコードになっていっていますね！
#### 追加で3件コメントいたしましたので、ご確認お願いいたします。

keeee-a committed:
修正2

resourcesのonly記述削除
privatemethodにredirect_to_indexとしてメソッド化
before_actionでログインしているユーザと出品しているユーザが同じかどうかの条件分岐をしているので、destroyメソッドのif @item.user.id == current_user.idという条件分岐を削除

chisesakai commented:
### LGTMです

レビュー依頼ありがとうございます！！
修正点確認させていただきました。
すっきりとまとまったコードになりましたね！素晴らしいです！

LGTMとさせていただきます。
実装お疲れ様でした。

今後の流れです。
マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。

引き続き、開発頑張っていきましょう！応援しております。
